### PR TITLE
Create layers based on a list of features.

### PR DESCRIPTION
**Pull request by manthey**
Currently, if you want to use a feature, you need to know what renderers in works in and manually check if those renderers are available on the current platform.  If the renderer isn't available, we fall back to the null renderer, which is only useful for the tile layers.

With this change, you can ask for a new layer based on the features you want.  If there are any renderers available that support all of the listed features, that renderer will be used.  Furthermore, there is now an inherent preference in renderers in the order vgl -> canvas -> d3 -> null, so we'll fall back to a better renderer than the null renderer in many instances.

Features can have capabilities.  For instance, the quad feature is only fully implemented in the vgl renderer.  Image quads are mostly implemented in the d3 and canvas renderers (when the quad is a parallelogram).  Color quads are implemented in the d3 renderer.  If you ask for a layer that supports {features: ['quad']}, you could get any of the vgl, d3, or canvas renderers (based on the support of that renderer).  If you ask for {features: ['quad.clr']}, you will only get either the vgl or d3 renderer.  If you ask for {features: ['quad.img-full']}, only vgl will do.

Many examples have been updated to use this.  This means that some examples will now work on more platforms (since they would only work if you have vgl, but now will work in other cases).

### CODE REVIEW

**Comment by aashish24** [comment]
🎉 will complete my review tomorrow but I liked what I reviewed so far

File: examples/choropleth/main.js [hunk 1]
```diff
@@ -31,11 +31,13 @@ $(function () {
...
-  // Create a gl feature layer
-  var vglLayer = map.createLayer(
+  // Create a feature layer.  We could either ask for a layer via a specific
+  // render {renderer: 'vgl'} or for a layer that supports our feature
+  // {features: ['choropleth']}.
+  var choroplethLayer = map.createLayer(
     'feature',
     {
-      renderer: 'vgl'
+      features: ['choropleth']
```
Comment (aashish24) [inline examples/choropleth/main.js]: +1

File: examples/hurricanes/main.js [hunk 2]
```diff
@@ -369,8 +369,8 @@ $(function () {
     }
   );
 
-  // Create a feature layer to draw on
-  layer = map.createLayer('feature');
+  // Create a feature layer to draw on.
+  layer = map.createLayer('feature', {features: ['line.multicolor']});
```
Comment (aashish24) [inline examples/hurricanes/main.js]: what is line.multicolor refers to?

Comment (aashish24) [inline examples/hurricanes/main.js]: I see now that you have added options at the time of registration. Interesting idea but now we have to make sure we are consistent about it and document it as well.

File: src/canvas/quadFeature.js [hunk 3]
```diff
@@ -143,5 +143,5 @@ var canvas_quadFeature = function (arg) {
 inherit(canvas_quadFeature, quadFeature);
 
 // Now register it
-registerFeature('canvas', 'quad', canvas_quadFeature);
+registerFeature('canvas', 'quad', canvas_quadFeature, {clr: false, img: true, 'img-full': false});
```
Comment (aashish24) [inline src/canvas/quadFeature.js]: could we change clr to color please? clr sometime could be confused with clear.

File: examples/hurricanes/main.js (hunk 2 above)
Comment (manthey) [inline examples/hurricanes/main.js]: Our gl lines can vary color from point to point.  The d3 lines can't.

We probably should document the feature capabilities in the base feature.

File: src/canvas/quadFeature.js (hunk 3 above)
Comment (aashish24) [inline src/canvas/quadFeature.js]: @manthey how do control and maintain this vocabulary? May we should make it part of public interface of baseclass (keys) and then the derived classes can just refer to it like this: 

registerFeature('canvas', 'quad', canvas_quadFeature, {quadFeature.color: false, quadFeature.img: true, quadFeature.img-full: false});

thoughts?

**Comment by aashish24** [comment]
Mostly looks great :yum:  to me but I have one suggestion for discussions.

File: src/canvas/quadFeature.js (hunk 3 above)
Comment (manthey) [inline src/canvas/quadFeature.js]: Sounds reasonable.  The base class can document what the keys mean, too.  I'll make those changes.

Comment (aashish24) [inline src/canvas/quadFeature.js]: great, thanks for the consideration @manthey

**Comment by aashish24** [comment]
LGTM :tada:

**Comment by aashish24** [comment]
@manthey should we go ahead and merge it?

**Comment by jbeezley** [comment]
LGTM

**Pull request merged**